### PR TITLE
altair: Add nft-sales pallet

### PR DESCRIPTION
**Pull request by NunoAlexandre**
We add the `nft_sales` pallet to `altair` and along with it we add `pallet_restricted_tokens`, `orml_tokens`, and
`pallet_permissions`, all pallets we need, directly or indirectly, for the `nft-sales`.

### CODE REVIEW

File: runtime/altair/src/lib.rs [hunk 1]
```diff
@@ -808,6 +820,155 @@ impl pallet_collator_selection::Config for Runtime {
...
+			send,
+			recv,
+			id,
+			amount: _amount,
+		} = details.clone();
+
+		match id {
+			CurrencyId::Usd | CurrencyId::Native => true,
+			CurrencyId::Tranche(pool_id, tranche_id) => {
+				P::has(pool_id, send, PoolRole::TrancheInvestor(tranche_id, UNION))
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs]: Done

File: pallets/permissions/src/weights.rs [hunk 2]
```diff
@@ -41,32 +41,32 @@ pub trait WeightInfo {
 pub struct SubstrateWeight<T>(PhantomData<T>);
 impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	fn add_as_admin() -> Weight {
-		(85_000_000 as Weight)
-			.saturating_add(T::DbWeight::get().reads(1 as Weight))
-			.saturating_add(T::DbWeight::get().writes(1 as Weight))
+		(14_000_000 as Weight)
```
Comment (mikiquantum) [inline pallets/permissions/src/weights.rs:44]: Did we run this on the bench instance on gcp?

File: runtime/altair/src/lib.rs [hunk 3]
```diff
@@ -1065,19 +1187,40 @@ impl_runtime_apis! {
...
+			list_benchmark!(list, extra, pallet_nft_sales, NftSales);
 
 			let storage_info = AllPalletsWithSystem::storage_info();
 
 			return (list, storage_info)
 		}
 	}
 }
+// Migration for scheduler pallet to move from a plain Call to a CallOrHash.
+pub struct SchedulerMigrationV3;
```
Comment (mikiquantum) [inline runtime/altair/src/lib.rs:1207]: This already happened, we can remove it

**Review by mikiquantum — COMMENTED** [review]
Lets pump up the runtime version and the node version as well.
Thanks!

File: pallets/permissions/src/weights.rs (hunk 2 above)
Comment (NunoAlexandre) [reply pallets/permissions/src/weights.rs:44]: is it to be run only on gcp?

File: runtime/altair/src/lib.rs (hunk 3 above)
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs:1207]: I am not sure why this was moved down here, but I will remove it then

File: runtime/altair/src/lib.rs [hunk 4]
```diff
@@ -808,6 +816,129 @@ impl pallet_collator_selection::Config for Runtime {
...
+	type PreFungibleMutateHold = common_traits::Always;
+	type PreFungibleTransfer = common_traits::Always;
+	type NativeFungible = Balances;
+	type NativeToken = NativeToken;
+	type WeightInfo = weights::pallet_restricted_tokens::SubstrateWeight<Self>;
+}
+
+parameter_types! {
+	pub ORMLMaxLocks: u32 = 2;
+}
```
Comment (mustermeiszer) [inline runtime/altair/src/lib.rs:905]: Might be a little too small?

File: runtime/altair/src/lib.rs [hunk 5]
```diff
@@ -856,9 +987,16 @@ construct_runtime!(
 		CrowdloanClaim: pallet_crowdloan_claim::{Pallet, Call, Storage, Event<T>, ValidateUnsigned} = 93,
 		CrowdloanReward: pallet_crowdloan_reward::{Pallet, Call, Storage, Event<T>} = 94,
 		CollatorAllowlist: pallet_collator_allowlist::{Pallet, Call, Storage, Config<T>, Event<T>} = 95,
+		Permissions: pallet_permissions::{Pallet, Call, Storage, Event<T>} = 96,
+		Tokens: pallet_restricted_tokens::{Pallet, Call, Event<T>} = 97,
```
Comment (mustermeiszer) [inline runtime/altair/src/lib.rs:991]: Do we want to expose call? Are we restricting this anyways in the filter of system?

File: runtime/altair/src/lib.rs (hunk 4 above)
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs:905]: These values are copied from `development`, I honestly don't know

File: runtime/altair/src/lib.rs [hunk 6]
```diff
@@ -808,6 +816,129 @@ impl pallet_collator_selection::Config for Runtime {
...
+}
+
+parameter_type_with_key! {
+	pub ExistentialDeposits: |currency_id: CurrencyId| -> Balance {
+		// every currency has a zero existential deposit
+		match currency_id {
+			_ => 0,
+		}
+	};
+}
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs:914]: @offerijns / @mikiquantum ?

File: runtime/altair/src/lib.rs (hunk 4 above)
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs:905]: @offerijns / @mikiquantum ?

File: pallets/permissions/src/weights.rs (hunk 2 above)
Comment (mikiquantum) [reply pallets/permissions/src/weights.rs:44]: yes we cannot use out own machines as benchmark env, instead lets use the cloud one that is following the recommended specs.

File: runtime/altair/src/lib.rs (hunk 5 above)
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs:991]: @offerijns / @mikiquantum ?

File: runtime/altair/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs]: I have updated this to

``` rust
pub ExistentialDeposits: |currency_id: CurrencyId| -> Balance {
	// every currency has a zero existential deposit
	match currency_id {
		CurrencyId::Native => ExistentialDeposit::get(),
		_ => 0,
	}
};
```

[the remark this answers is not in the document]
Comment (offerijns) [reply runtime/altair/src/lib.rs]: I think re-using the existential deposit defined in the balances pallet for the tokens pallet makes sense!

File: pallets/permissions/src/weights.rs (hunk 2 above)
Comment (NunoAlexandre) [reply pallets/permissions/src/weights.rs:44]: We are running all the benchmarks in a follow up PR when all the pallets are taken into account.

File: runtime/altair/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply runtime/altair/src/lib.rs]: @mikiquantum will wait for your take here

**Comment by NunoAlexandre** [comment]
Everything is addressed and I reverted the changes to existing weights to avoid a conflict with the upcoming changes in https://github.com/centrifuge/centrifuge-chain/pull/689.

We will bump the altair runtime and node versions once we are ready to cut out the new release.

**Pull request merged**